### ISSUE TITLE
Make VersionByNamespaceConvention public available symbol

### ISSUE THREAD

**Issue by juanmanuelrios** [dotnet/aspnet-api-versioning#741]
It would be very useful to have VersionByNamespaceConvention (and any other convention to determine API version as well) be public available symbol, to be used anywhere in the code.

It might be possible to abstract away the code from the partial class where it is placed now, and make it available by providing them through some set of abstractions / contract interface.

By having conventions publicly available, one can build features such as HttpClient's routing based on the same framework's scaffolding used to get the version number from. 

It's worth saying that unit-testing can take advantage of it as well.

thanks in advance.

**Comment by commonsensesoftware** [comment]
It's not entirely clear what you are asking for. **VersionByNamespaceConvention** is `public` and is the only out-of-the-box provided convention. The **IControllerConvention** interface is also `public`, which is how you apply custom conventions. The entire _conventions_ set of APIs are public and customizable as well.

The **IApiVersionParameterSource** interface describes _where_ an API parameter comes from. The **IApiVersionReader** must implement this interface as well. All out-of-the-box API version reader implementations provide this functionality. This capability is used in the API Explorer extensions to _describe_ API version parameters.

You should be aware that API version readers and sources can be composed. It is possible, and perfectly valid, to have a composite reader that can read from multiple sources such as HTTP header and query string. This can make it challenging for clients or Open API (formerly Swagger) document generation. You may only need one, but multiple are allowed. This means you'll have to pick one.

It may be worth clarifying that _conventions_ are how API versions are applied as metadata and **not** how API versions are determined/evaluated at runtime. The **IApiVersionReader** is the contract to read an API version from a HTTP request. The **IApiVersionParameterSource** describes information about that parameter (e.g. location, description, default, value, etc).

Using the collation features of the API version-aware API Explorer extensions are quite interesting for testing scenarios, even though Open API document generation tends to be one the most common use case. API Versioning doesn't care about nor dependent on Open API. Described APIs are collated by version, which you can use to drive test runs.

**Comment by juanmanuelrios** [comment]
Thanks for your early response, and sorry for the delayed mine. I'll expand my statement, because it's a little misleading as it was stated.

We've built an application service abstraction layer on top of MVC scaffolding to expose application services as if they were API controllers, using conventional routing to wire up their methods to the routing  sub-system. 

Our framework discover certain tagged types (those with an special interface tag) and wires up the endpoint routing to get their methods accessible through http. They behave as if they were really api controllers, from the standpoint of the consumer.

Here is where VersionByNamespaceConvention comes in, because versioning is applied to said abstraction layer during routing at server-side, the endpoint route considers the application service's namespace. This is ok, no issues so far.

The problem is found client-side, because we also built an application service's proxy to create clients that proxify calls to the server, leveraging DispatchProxy and HttpClient features. DispatchProxy takes the application service's contract and, by using the same conventions we've used server-side, we expect to reach the target endpoint (server-side) by building the invocation route and pass it to the HttpClient instance, to get the invocation job done.  Client-side we are not able to get the same routing information as server-side because the version extracting logic is placed in a non mvc-independent symbol. This is where I'm point at.

**Comment by commonsensesoftware** [comment]
I see what you're trying to do. A small handful of people have made somewhat similar requests in the past. One of the most frequently asked for pieces is the **ApiVersion** so it can be used in clients. I can see it making sense to move some of these core bits into an _Abstractions_ base library. The work isn't all that much, but there is some considerable thought that needs to go into that and it would require a major version bump across the board. I'll track this for future feature consideration.

Your particular request would also require splitting apart the parsing parts from the API conventions part. Currently, they are intertwined. Client-side has never been a goal of this project, but I'm not opposed to helping people provide such extensions. In general, the server cannot make assumptions about what a client wants - unless you own both sides, which is your case.

This is a reasonable feature request, but it will likely take some time to get out. My recommendation to unblock you, if you haven't already, is the fork the relevant bits from [VersionByNamespaceConvention.cs](../blob/master/src/Common/Versioning/Conventions/VersionByNamespaceConvention.cs). The only thing you really need is the regular expression. The implementation has to cover all possible scenarios, which is likely much more complex and involved than your specific scenario. I would only take the bits that you need. If you're worried about parity, you can also take the [test suite](../blob/master/test/Microsoft.AspNetCore.Mvc.Versioning.Tests/Versioning/Conventions/VersionByNamespaceConventionTest.cs). 

The only other option would be hiding the implementation behind DI. That doesn't completely remove the MVC dependency at runtime, but it would from your core libraries.

**Comment by leonardochaia** [comment]
Hey @commonsensesoftware , we (I work with @juanmanuelrios ) have kinda followed the approach you recommended. Basically we created an abstraction that allows us to get the version for `System.Type`.
We then implemented it basically copying from VersionByNamespaceConvention.
We use the abstraction in our internal framework on the client, and on the server we also implemented `IControllerConvention` to use our abstraction, just in case it diverges from what you guys provide.

That being said, we agree is a valid feature request and it would be ideal to be able to abstract a way this so that enterprise frameworks that build on top of MVC can take advantage of it, as I understand this is also one of the project goals.

> Client-side has never been a goal of this project, but I'm not opposed to helping people provide such extensions. In general, the server cannot make assumptions about what a client wants - unless you own both sides, which is your case.

This is definitively our case.

So just wanted to comment what we did and say thanks for tracking this as an enhancement!

**Comment by juanmanuelrios** [comment]
Hi guys.

As Leo said, we've already taken the path of copying some useful pieces of
code from VersionByNameSpaceConvention, and it fulfilled our requirements
so far. We'd really appreciate your effort toward isolating these pieces
(needed client-side) from MVC, (server-side) so we would build client side
features expecting they behave the same way.

Thanks a lot for your effort, you are doing an excellent job supporting the
community.

**Comment by commonsensesoftware** [comment]
This feature and capability is now supported in `6.0`. It is available in [Asp.Versioning.Abstractions](https://www.nuget.org/packages/Asp.Versioning.Abstractions). Please make sure you've read all the announcements and release notes; there's a lot of changes.

This is exposed via the `NamespaceParser` which builds a read-only list of API versions from a namespace. Note that it is possible to pick up multiple API versions from a single namespace. That _should_ be considered ambiguous, but it's up to you to handle. The parser does only one thing - parse. The type is also extensible if you require further customization. If not, you can simply use `NamespaceParser.Default`.

There is also now [Asp.Versioning.Http.Client](https://www.nuget.org/packages/Asp.Versioning.Http.Client) which enables making `HttpClient` API version-aware. I'm not entirely sure how namespaces can be uses on the client-side - yet. If you have some ideas, please open a new issue for that feature and we can track it. Thanks.
